platforms/nuttx/src/px4/common/px4_init.cpp

```diff
@@ -44,6 +44,94 @@
 
 #include <fcntl.h>
 
+#if defined(GPIO_OTGFS_VBUS) && defined(CONFIG_SYSTEM_CDCACM)
+__BEGIN_DECLS
+#include <nuttx/wqueue.h>
+#include <builtin/builtin.h>
+
+extern int sercon_main(int c, char **argv);
+extern int serdis_main(int c, char **argv);
+__END_DECLS
+
+#include <uORB/Subscription.hpp>
+#include <uORB/topics/actuator_armed.h>
+
+static struct work_s usb_serial_work;
+static int vbus_value_prev = 0;
+static int vbus_connected_count = 0;
+static bool sercon_running = false;
+static bool mavlink_usb_started = false;
```

Comment:
Would this all be simpler with a state machine?
enum state_t {conecting, connected, disconnecting, disconnected};
state_t state = disconnected;
static uint32_t input = 0x0;
input <<= 1;
input |= px4_arch_gpioread(GPIO_OTGFS_VBUS)  
input &= 0x3;
if (input == 0 && state == connected) run the disconnect  sequence
or input (==3 && state == disconnected) run the connect sequence.

Reply:
Thanks I'll give it a try.

Reply:
Done.